modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C

```diff
@@ -0,0 +1,210 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ApplyPenetrationConstraintLMMechanicalContact.h"
+#include "PenetrationLocator.h"
+#include "PenetrationInfo.h"
+#include "SystemBase.h"
+#include "Assembly.h"
+
+#include "DualRealOps.h"
+
+using MetaPhysicL::DualNumber;
+
+registerMooseObject("ContactApp", ApplyPenetrationConstraintLMMechanicalContact);
+
+InputParameters
+ApplyPenetrationConstraintLMMechanicalContact::validParams()
+{
+  InputParameters params = NodeFaceConstraint::validParams();
+  params.set<bool>("use_displaced_mesh") = true;
+
+  params.addCoupledVar("disp_y", "The y displacement");
+  params.addCoupledVar("disp_z", "The z displacement");
+  params.addParam<Real>("c", 1, "Parameter for balancing the size of the gap and contact pressure");
+
+  params.addClassDescription(
+      "Implements the KKT conditions for normal contact using an NCP "
+      "function, in this case just the min function. This function enforces that either "
+      "the gap distance or the normal contact pressure (represented by the value of `variable`) is "
+      "zero.");
+  return params;
+}
+
+ApplyPenetrationConstraintLMMechanicalContact::ApplyPenetrationConstraintLMMechanicalContact(
+    const InputParameters & parameters)
+  : NodeFaceConstraint(parameters),
+    _c(getParam<Real>("c")),
+    _fe_problem(*getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
+    _displaced_problem(_fe_problem.getDisplacedProblem().get()),
+    _residual_copy(_sys.residualGhosted())
+{
+}
+
+bool
+ApplyPenetrationConstraintLMMechanicalContact::shouldApply()
+{
+  const bool has_dofs = _current_node->n_dofs(_sys.number(), _var.number());
```

Comment:
If the LM doesn't have any dofs at the node, e.g. a first order LM on a mid-edge node, then we do not apply the constraint because there is no dof to constrain